warehouse/query-core/src/main/java/datawave/next/DocIdIteratorVisitor.java

```diff
@@ -0,0 +1,521 @@
+                if (result == null) {
+                    result = scanResult;
+                } else {
+                    result.addKeys(scanResult.getResults());
+                }
+            } else {
+                if (log.isTraceEnabled()) {
+                    log.trace("Node did not return a set: {}", JexlStringBuildingVisitor.buildQuery(child));
+                }
+            }
+        }
+
+        if (result == null) {
+            // no term was executable
+            if (log.isTraceEnabled()) {
+                log.trace("union: [{}] found 0 hits", JexlStringBuildingVisitor.buildQuery(node));
+            }
+            return data;
+        }
+
+        if (log.isTraceEnabled()) {
+            log.trace("union: [{}] found {} hits", JexlStringBuildingVisitor.buildQuery(node), result.getResults().size());
+        }
+        return result;
+    }
+
+    /*
+     * There are many potential types of joins happening here. Enumerate and work through cases.
+     */
+
+    @Override
+    public Object visit(ASTAndNode node, Object data) {
+        QueryPropertyMarker.Instance instance = QueryPropertyMarker.findInstance(node);
+        if (instance.isAnyType()) {
+            // always pass context to a marker node. If the parent is a union, the only way context was passed in is because context exists from a grandparent
+            // intersection as in the case of (A and (B or marker))
+            return handleMarker(node, data, instance);
+        }
+
+        List<JexlNode> positive = new ArrayList<>();
+        List<JexlNode> negative = new ArrayList<>();
+        for (int i = 0; i < node.jjtGetNumChildren(); i++) {
+            JexlNode deref = JexlASTHelper.dereference(node.jjtGetChild(i));
+            if (deref instanceof ASTNotNode) {
+                negative.add(deref);
+            } else {
+                positive.add(deref);
+            }
+        }
+
+        // positive terms first
+        ScanResult result = null;
+        for (JexlNode child : positive) {
+            // intersections drive their own context
+            Object o = child.jjtAccept(this, result);
+            if (!(o instanceof ScanResult)) {
+                if (log.isDebugEnabled()) {
+                    log.debug("Node did not return a set: {}", JexlStringBuildingVisitor.buildQuery(child));
+                }
+                continue;
+            }
+
+            ScanResult scanResult = (ScanResult) o;
+            if (scanResult.getResults().isEmpty()) {
+                if (log.isDebugEnabled()) {
+                    log.debug("short circuit intersection, child returned zero hits");
+                }
+                return new HashSet<>();
+            }
+
+            if (result == null) {
+                result = scanResult;
+            } else if (scanResult.isTimeout()) {
+                log.debug("scan timed out");
+                // TODO: partial intersection with sorted sets
```

Comment:
probably need to do something here?

Reply:
Created https://github.com/NationalSecurityAgency/datawave/issues/2914